### ISSUE TITLE
Train and predict on same the device for `skorch.NeuralNet`

### ISSUE THREAD

**Issue by kqf** [skorch-dev/skorch#706]
Hi, 

I have a bit silly question about `skorch.NeuralNet` inference. As I see `skorch.NeuralNet.predict_proba` method will run on CPU regardless of the `device` parameter passed to the `skorch.NeuralNet.__init__` method:

1. `skorch.NeuralNet.predict_proba` calls ` self.forward_iter(X, training=False)` (see [here](https://github.com/skorch-dev/skorch/blob/master/skorch/net.py#L1122))
2. The default device for this method is [CPU](https://github.com/skorch-dev/skorch/blob/master/skorch/net.py#L927)

As a workaround I use this code:

```python
import skorch


class MyNeuralNet(skorch.NeuralNet):
    def forward_iter(self, X, training=False, device="cpu"):
        return super().forward_iter(X, training=training, device=self.device)


```
This looks unnatural. Is there a better way to achieve this? Perhaps I overlooked something. This question seems even more puzzling as I stumbled upon issue #553 which reports an opposite problem.

**Comment by BenjaminBossan** [comment]
Hi, it's not a silly question.

IMO it should work correctly. See here:

https://github.com/skorch-dev/skorch/blob/f34a885a4e978b25da4a904dd11fceb20816e83a/skorch/net.py#L964-L969

The `device` parameter is not used for the inference step (`self.evaluation_step(Xi, training=training)`), it is only used to cast the prediction (`yield to_device(yp, device=device)`). The idea is that you can control whether the final prediction will reside on CPU or GPU. Often, you want the former, which is the default, because you don't want the prediction to clog up your GPU memory.

Did you actually observe that inference happens on CPU despite the device being cuda or was it only from the reading of the code? Because if you actually observed it, we need to investigate further.

**Comment by kqf** [comment]
Indeed, you are right. I misidentified the problematic spot. The prediction is on the same device as desired. In my case I need to take only `k` logits from the output to make the final prediction. I used `predict_nonlinearity` method to achieve that. Here is the code to reproduce the error:

```python
import torch
import skorch


def dataloader(x, batch_size, n_items, n_labels, *args, **kwargs):
    for i in x:
        X = torch.rand(batch_size, n_items)
        y = torch.randint(0, n_labels, (batch_size,))
        yield X, y


class MyModule(torch.nn.Module):
    def __init__(self, n_items, n_labels):
        super().__init__()
        self._fc = torch.nn.Linear(n_items, n_labels)

    def forward(self, X):
        assert X.device == torch.device("cuda:0"), "this will never fail"
        return self._fc(X)


def topk(logits):
    # The problematic place
    assert logits.device == torch.device("cuda:0")
    return torch.topk(logits, -1)[1].detach().cpu()


model = skorch.NeuralNet(
    module=MyModule,
    module__n_items=100,
    module__n_labels=10,
    predict_nonlinearity=topk,  # << This is an issue
    device=torch.device("cuda:0" if torch.cuda.is_available() else "cpu"),
    iterator_train=dataloader,
    iterator_train__n_items=100,
    iterator_train__n_labels=10,
    iterator_valid=dataloader,
    iterator_valid__n_items=100,
    iterator_valid__n_labels=10,
    criterion=torch.nn.CrossEntropyLoss,
)
model.fit(range(5))
model.predict(range(2))

```
Thanks for explanation now it's more clear what is going on. So, perhaps the right question to ask is how to force outputs to be stored on the same device?

**Comment by BenjaminBossan** [comment]
I think there is not really anything special that needs to be done, or maybe I misunderstand?

```python
device = ...  # define your desired device
net = NeuralNet(..., device=device)
y_fwd = net.forward(X, device=device)
```

**Comment by kqf** [comment]
Oh, 😮.  I was missing that point, thanks a lot. It works for me even without any parameters:
```python
y_fwd = net.forward(X)
```
This resolves my issue. Thank you so much, and sorry for the noise
